### PR TITLE
Use prefix store for contract history

### LINKED ISSUE #196: Move history of contract code migrations to their own prefix store.
## Summary

Move history of contract code migrations to their own prefix store. This would reduce the overhead in de-/serialization data for ContractInfo and save gas.

Follow up on #130 
____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by alpe**
Resolves #196 
After #199 

Includes new CLI command: 
`wasmcli query wasm contract-history [bech32_address]` to print all the code changes.
```json
wasmcli q wasm contract-history cosmos18r5szma8hm93pvx6lwpjwyxruw27e0k5uw835c
[
  {
    "operation": "Init",
    "code_id": 1
  },
  {
    "operation": "Migrate",
    "code_id": 2
  }
]
```

Note: to be on the safe side sequences have their own prefix as well.

**This PR also updates the codec to fix the message decoding issue.**
______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/wasm/internal/keeper/genesis_test.go [hunk 1]
```diff
@@ -482,14 +445,16 @@ func TestImportContractWithCodeHistoryReset(t *testing.T) {
...
-		ContractCodeHistory: []types.ContractCodeHistoryEntry{{
-			Operation: types.GenesisContractCodeHistoryType,
-			CodeID:    1,
-			Updated:   types.NewAbsoluteTxPosition(ctx),
-		},
-		},
 	}
 	assert.Equal(t, expContractInfo, *gotContractInfo)
+
+	expHistory := []types.ContractCodeHistoryEntry{{
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/genesis_test.go:451]: :+1:

File: x/wasm/internal/keeper/keeper.go [hunk 2]
```diff
@@ -376,6 +377,27 @@ func (k Keeper) setContractAdmin(ctx sdk.Context, contractAddress, caller, newAd
...
 
+func (k Keeper) appendToContractHistory(ctx sdk.Context, contractAddr sdk.AccAddress, newEntries ...types.ContractCodeHistoryEntry) {
+	prefixStore := prefix.NewStore(ctx.KVStore(k.storeKey), types.ContractHistoryStorePrefix)
+	var entries []types.ContractCodeHistoryEntry
+	bz := prefixStore.Get(contractAddr)
+	if bz != nil {
+		k.cdc.MustUnmarshalBinaryBare(bz, &entries)
+	}
+	entries = append(entries, newEntries...)
+	prefixStore.Set(contractAddr, k.cdc.MustMarshalBinaryBare(&entries))
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/keeper.go:388]: Couldn't you simplify this a bit?

```go
  entries = append(k.GetContractHistory(ctx, contractAddr), newEntries...)
  prefixStore := prefix.NewStore(ctx.KVStore(k.storeKey), types.ContractHistoryStorePrefix)
  prefixStore.Set(contractAddr, k.cdc.MustMarshalBinaryBare(&entries))
```

File: x/wasm/internal/keeper/querier_test.go [hunk 3]
```diff
@@ -214,9 +215,102 @@ func TestListContractByCodeOrdering(t *testing.T) {
...
+				CodeID:    1,
+				Updated:   types.NewAbsoluteTxPosition(ctx),
+				Msg:       []byte(`"init message"`),
+			}},
+			expContent: []types.ContractCodeHistoryEntry{{
+				Operation: types.GenesisContractCodeHistoryType,
+				CodeID:    1,
+			}},
+		},
+		"response with multiple entries": {
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/querier_test.go:249]: :+1:

**Comment by alpe** [comment]
Pushed a fix for the history query response together with a `api_migration.md` file

File: api_migration.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
+# Changes to the api
```
Comment (ethanfrey) [inline api_migration.md:1]: Thank you for adding this.

There is also a code -> code_id renaming from another PR I believe

Comment (alpe) [reply api_migration.md:1]: It was `code_id` in the json before. Only the Go code was inconsistent. There are no changes for the clients

**Pull request merged**